File: js/src/bowling_game.js
var BowlingGame = function () {
  var rolls = [];

  function roll(pins) {
    rolls.push(pins);
    if(pins == 10) rolls.push(0);
  }

  function score() {
    var score = 0;
    for (var frame = 1; frame <= 10; frame++) {
      score += scoreFor(frame);
    }
    return score;
  }

  function scoreFor(frame) {
    if(isStrike(frame))
      return standardScoreOf(frame) + strikeBonusFor(frame);
    if(isSpare(frame))
      return standardScoreOf(frame) + spareBonusFor(frame);
    return standardScoreOf(frame);
  }

  function isStrike(frame) {
    return firstRollOf(frame) == 10;
  }

  function strikeBonusFor(frame) {
    return firstRollOf(frame + 1) + secondRollOf(frame + 1);
  }

  function isSpare(frame) {
    return standardScoreOf(frame) == 10;
  }

  function spareBonusFor(frame) {
    return firstRollOf(frame + 1);
  }

  function standardScoreOf(frame) {
    return firstRollOf(frame) + secondRollOf(frame);
  }

  function firstRollOf(frame) {
    return rollAt((frame - 1) * 2);
  }

  function secondRollOf(frame) {
    return rollAt((frame - 1) * 2 + 1);
  }

  function rollAt(index) {
    return rolls[index] || 0;
  }

  return {roll: roll, score: score};
};

File: csharp/XPPractices/BowlingGame.cs
using System.Collections.Generic;

namespace XPPractices
{
    public class BowlingGame : IBowlingGame
    {
        private readonly List<int> rolls = new List<int>();
        private bool isFirstRoll = true;

        public void Roll(int pins)
        {
            rolls.Add(pins);
            if (pins == 10 && isFirstRoll)
                rolls.Add(0); // dirty hack to keep the rolls in pairs
            else
                isFirstRoll = !isFirstRoll;
        }

        public int GetScore()
        {
            var score = 0;
            for (var frame = 1; frame <= 10; frame++)
                score += ScoreFor(frame);
            return score;
        }

        private int ScoreFor(int frame)
        {
            if (IsStrike(frame)) return NormalScoreFor(frame) + StrikeBonusFor(frame);
            if (IsSpare(frame))
                return NormalScoreFor(frame) + SpareBonusFor(frame);
            return NormalScoreFor(frame);
        }

        private bool IsStrike(int frame)
        {
            return FirstRollOf(frame) == 10;
        }

        private int StrikeBonusFor(int frame)
        {
            if (IsStrike(frame + 1))
            {
                return FirstRollOf(frame + 1) + FirstRollOf(frame + 2);
            }
            return FirstRollOf(frame + 1) + SecondRollOf(frame + 1);
        }

        private int SpareBonusFor(int frame)
        {
            return FirstRollOf(frame + 1);
        }

        private int NormalScoreFor(int frame)
        {
            return FirstRollOf(frame) + SecondRollOf(frame);
        }

        private bool IsSpare(int frame)
        {
            return FirstRollOf(frame) + SecondRollOf(frame) == 10;
        }

        private int FirstRollOf(int frame)
        {
            return RollAt((frame - 1) * 2);
        }

        private int SecondRollOf(int frame)
        {
            return RollAt((frame - 1) * 2 + 1);
        }

        private int RollAt(int index)
        {
            return index < rolls.Count ? rolls[index] : 0;
        }
    }
}